### PR TITLE
Enable building with dotnetcli for non-Windows

### LINKED ISSUE #652: Build against .Net Core and run tests on Linux
By design, PSReadLine builds against .Net Desktop to minimize the number of dlls that need to be included in the module, but it should be possible to build against .Net Core.

This would be useful for 2 reasons - development on non-Windows platforms, and adding CI for non-Windows platforms.

The unit tests already use xUnit, so hopefully it won't be too hard to get them running on Linux.

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
Ported csproj to use dotnetcli.  When run from PSCore6, builds for netcoreapp21.  When run from Windows PowerShell 5.1, builds for net461.  Tests ran on Windows PowerShell, PSCore6 Windows, PSCore6 Ubuntu 18.04, and PSCore6 macOS all passing.  Linux run requires xclip to be installed as that is necessary for clipboard to function (which tests depend upon).  Using dotnetsdk 2.1.  Appropriate .Net Framework 4.6.1 SDK is required for Windows PowerShell build.

Fix #652

### CODE REVIEW

**Comment by lzybkr** [comment]
You should add a new task in `PSReadLine.build.ps1` instead of adding the new file `build.ps1`.

The new task should probably be invoked from `appveyor.yml`.

And I would much rather see changes to `PSReadLine\PSReadLine.csproj` - I'm guessing there are about 2 lines of real difference between the files.

**Comment by SteveL-MSFT** [comment]
@lzybkr I'm fine having a single csproj, but wasn't sure if you preferred keeping the existing project as-is.

**Comment by lzybkr** [comment]
Does building against PowerShell Standard work against Windows PowerShell 5.1 on Win7 and Win10 without shipping type forwarding dlls? If so, they fine, but I'm guessing there isn't a clean solution to that problem yet.

A single csproj can build multiple ways with conditions though - and that's what I was suggesting.

**Comment by SteveL-MSFT** [comment]
@lzybkr If Win7 doesn't have .Net 4.7, the shims are needed.  Newer Win10 has 4.7.x.  We plan to have a module that contains the .Net Std shims that other modules can reference.

**Comment by SteveL-MSFT** [comment]
@lzybkr are you ok if we move both the Windows and non-Windows build to dotnetcli?  Would simplify things quite a bit.

**Comment by lzybkr** [comment]
@SteveL-MSFT - if you submit a PR so I can see how it's simpler, that would help.

I don't want to distribute unnecessary private copies of dlls, I'm already unhappy about needing `System.Runtime.InteropServices.RuntimeInformation.dll`.

I also don't want runtime logic that might affect startup time, e.g. path searching for type forwarders dlls.

**Comment by SteveL-MSFT** [comment]
I'm going to have to put this on hold for now.  Got it building and tests running with dotnetcli, but tests fail to run correctly.

**Comment by SteveL-MSFT** [comment]
Tests pass on macOS and Windows w/ PSCore6. Out of town next few days. Will get it working in WinPS after I get back

**Comment by SteveL-MSFT** [comment]
@lzybkr I think this is ready for review

File: PSReadLine.build.ps1 [hunk 1]
```diff
@@ -139,46 +157,48 @@ $buildTestParams = @{
...
-        $runner = "$PSScriptRoot\PSReadLine\packages\xunit.runner.console.2.3.1\tools\net452\xunit.console.exe"
-        if ($env:APPVEYOR)
-        {
-            $outXml = "$PSScriptRoot\xunit-results.xml"
-            & $runner $PSScriptRoot\test\bin\$Configuration\PSReadLine.Tests.dll -appveyor -xml $outXml
-            $wc = New-Object 'System.Net.WebClient'
-            $wc.UploadFile("https://ci.appveyor.com/api/testresults/xunit/$($env:APPVEYOR_JOB_ID)", $outXml)
-        }
-        else
+task RunTests {
```
Comment (lzybkr) [inline PSReadLine.build.ps1]: You removed `BuildTests` - please put in the dependency so one can run `Invoke-Build Clean,RunTests` .

File: PSReadLine.build.ps1 [hunk 2]
```diff
@@ -199,15 +219,26 @@ task LayoutModule BuildMainModule, BuildMamlHelp, {
         Copy-Item $file $targetDir
     }
 
-
-    Copy-Item PSReadLine/bin/$Configuration/Microsoft.PowerShell.PSReadLine2.dll $targetDir
-    Copy-Item PSReadLine/bin/$Configuration/System.Runtime.InteropServices.RuntimeInformation.dll $targetDir
```
Comment (lzybkr) [inline PSReadLine.build.ps1:204]: Why did you remove copying `System.Runtime.InteropServices.RuntimeInformation.dll`? Is this dll installed everywhere PSReadLine including Win7?

**Review by lzybkr — CHANGES_REQUESTED** [review]
I think you should add instructions for building to `README.md` so I can easily point people to what they need to install.

You also need to try running the tests from VS. You did not update the `TestPSReadLine` project and that fails to build, preventing tests from running. (That project is used for interactive development/testing without affecting PowerShell by installing the module.)

File: PSReadLine.build.ps1 [hunk 3]
```diff
@@ -139,46 +157,48 @@ $buildTestParams = @{
...
-            $wc.UploadFile("https://ci.appveyor.com/api/testresults/xunit/$($env:APPVEYOR_JOB_ID)", $outXml)
-        }
-        else
+task RunTests {
+    $env:PSREADLINE_TESTRUN = 1
+    $runner = $script:dotnet
+
+    if ($PSVersionTable.PSEdition -eq "Core")
+    {
+        $psAssemblies = "System.Management.Automation.dll", "Newtonsoft.Json.dll", "System.Management.dll", "System.DirectoryServices.dll"
```
[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply PSReadLine.build.ps1:171]: The new SDK we published should remove the need for this line, it was broken previously

File: PSReadLine/Clipboard.cs [hunk 4]
```diff
@@ -3,38 +3,154 @@
...
+                {
+                    _clipboardSupported = false;
+                    PSConsoleReadLine.Ding();
+                    return "";
+                }
+
+                clipboardText = process.StandardOutput.ReadToEnd();
+                process.WaitForExit(250);
+
+                if (process.ExitCode != 0)
```
[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply PSReadLine/Clipboard.cs]: will fix

**Comment by SteveL-MSFT** [comment]
@lzybkr verified solution builds and tests run in VS2017

File: PSReadLine.build.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (lzybkr) [reply PSReadLine.build.ps1:171]: OK - we'll put off using `-runtime` until PowerShellGet supports installing just the necessary binaries for a given install.

**Review by lzybkr — DISMISSED** [review]

**Comment by SteveL-MSFT** [comment]
@lzybkr the .psd1 file wasn't changed in this PR and looking in my repo, it's still set to 2.0.0

**Review by lzybkr — APPROVED** [review]

**Pull request merged**